### PR TITLE
Fix json marshal/unmarshal implementations for SCRAM messages

### PR DESCRIPTION
**Pull request by yulicrunchy**
Also add json marshaling for AuthenticationOK.

### CODE REVIEW

**Comment by yulicrunchy** [comment]
@henvic @jackc ptal

File: authentication_sasl.go [hunk 1]
```diff
@@ -58,3 +59,27 @@ func (src *AuthenticationSASL) Encode(dst []byte) []byte {
...
+		Type           string
+		AuthMechanisms []string
+	}{
+		Type:           "AuthenticationSASL",
+		AuthMechanisms: src.AuthMechanisms,
+	})
+}
+
+// UnmarshalJSON implements encoding/json.Unmarshaler.
+func (dst *AuthenticationSASL) UnmarshalJSON(data []byte) error {
```
Comment (henvic) [inline authentication_sasl.go:75]: Why did you implement this function? Isn't it going to do the exact same if it fallbacks to Go's default unmarshaling given the Type is never used?

File: json_test.go [hunk 2]
```diff
@@ -23,9 +23,9 @@ func TestJSONUnmarshalAuthenticationMD5Password(t *testing.T) {
 }
 
 func TestJSONUnmarshalAuthenticationSASL(t *testing.T) {
-	data := []byte(`{"Type":"AuthenticationSASL", "AuthMechanisms":[]}`)
+	data := []byte(`{"Type":"AuthenticationSASL","AuthMechanisms":["SCRAM-SHA-256"]}`)
```
Comment (henvic) [inline json_test.go:26]: Would only having this test without the Unmarshal implementation code already be enough?

File: authentication_sasl.go [hunk 3]
```diff
@@ -58,3 +59,27 @@ func (src *AuthenticationSASL) Encode(dst []byte) []byte {
 
 	return dst
 }
+
+// MarshalJSON implements encoding/json.Marshaler.
+func (src AuthenticationSASL) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		Type           string
+		AuthMechanisms []string
```
Comment (henvic) [inline authentication_sasl.go:67]: Maybe embedding is better?

```suggestion
		Type           string
		AuthenticationSASL
```

**Comment by henvic** [comment]
Hi @yulicrunchy,

I've noticed that a JSON Marshaler is also missing for the following structs:

* AuthenticationCleartextPassword
* AuthenticationMD5Password
* ErrorResponse
* NoticeResponse (which is a `type ErrorResponse`, so solving for ErrorResponse is enough)

File: authentication_sasl.go (hunk 1 above)
Comment (yulicrunchy) [reply authentication_sasl.go:75]: With the implementation:

```
B{"Type":"AuthenticationSASL","AuthMechanisms":["SCRAM-SHA-256"]}
```

Without:

```
B{"AuthMechanisms":["SCRAM-SHA-256"]}
```

Comment (yulicrunchy) [reply authentication_sasl.go:75]: thanks, will delete.

File: json_test.go (hunk 2 above)
Comment (yulicrunchy) [reply json_test.go:26]: Yes, it works without the custom implementation, good catch.

File: authentication_sasl.go (hunk 3 above)
Comment (yulicrunchy) [reply authentication_sasl.go:67]: I don't think this will work since it will lead to infinite recursion when marshaling itself.

**Comment by yulicrunchy** [comment]
I added the missing implementations as requested.

**Comment by jackc** [comment]
LGTM

**Comment by henvic** [comment]
Hey @jackc, could you please tag a new version? Thank you!

**Pull request merged**